File: src/main/java/ca/frar/jjjrmi/translator/EncodedArray.java
package ca.frar.jjjrmi.translator;
import ca.frar.jjjrmi.exceptions.JJJRMIException;
import java.lang.reflect.Array;
import org.json.JSONArray;
import org.json.JSONObject;

class EncodedArray extends JSONObject{
    private final TranslatorResult encodedResult;
    private Object object = null;
    private final JSONArray elements;

    EncodedArray(Object object, TranslatorResult encodedResult) throws JJJRMIException {
        this.encodedResult = encodedResult;
        this.object = object;
        this.elements = new JSONArray();        
        this.put(Constants.KeyParam, encodedResult.getTranslator().allocReference(object));
        this.put(Constants.RetainParam, false);
        this.put(Constants.SizeParam, Array.getLength(object));
        encodedResult.getTranslator().addTempReference(this.get(Constants.KeyParam).toString(), object);
        this.put(Constants.ElementsParam, elements);
        encode();
    }

    private void encode() throws JJJRMIException {
        for (int i = 0; i < (Array.getLength(object)); i++) {
            Object element = Array.get(object, i);
            elements.put(new Encoder(element, encodedResult).encode());
        }
    }
}

File: src/main/js/translator/EncodedArray.js
"use strict";
const Array = require("./Array");
const Constants = require("./Constants");
const JSONArray = require("./JSONArray");
class EncodedArray {
	constructor(object, encodedResult) {
            this.encodedResult = encodedResult;
            this.object = object;
            this.elements = [];
            this[Constants.KeyParam] = encodedResult.getTranslator().allocReference(object);
            this[Constants.RetainParam] = false;
            encodedResult.getTranslator().addTempReference(this[Constants.KeyParam], object);
            this[Constants.ElementsParam] = this.elements;
            this.encode();
        }
	encode() {
            for(let i = 0; i < this.object.length; i++){
                let element = this.object[i];
                this.elements[i] = new Encoder(element, this.encodedResult).encode();
            }
        }
};

if (typeof module !== "undefined") module.exports = EncodedArray;